### PR TITLE
Plot window

### LINKED ISSUE #9: Simple Plot
- tkinter
- plotly

### PR DESCRIPTION
**Pull request by FrnRmn**
- Added GUI with a live line plot

If you have suggestions on graph properties to change, they are welcome.
Currently, some graph parameters are hardcoded in the Co2Cpu.py module, maybe we can export them in a config.yaml file.

### CODE REVIEW

File: src/core/cpuCO2.py [hunk 1]
```diff
@@ -26,39 +29,83 @@ def cpu_co2(country_code = None) -> None:
...
         print(f'\nRetrieving CPU usage of the next {cpu_retrieval_time} seconds...')
         cpu_info = cpu.get_cpu_info()
         cpu_usage = cpu.get_cpu_usage(cpu_retrieval_time) #it seems too low to be true!
         print("- Current CPU: {0} \n- CPU usage during last 10s: {1}%".format(cpu_info, cpu_usage))
 
         # Combine Co2 with CPU usage
         print('\nCombining Co2 and CPU data...')
-        co2_emissions = combine_cpu_CO2(cpu_usage, cpu.get_cpu_tdp(cpu_info), carbon_intensity)
+        co2_emissions = combine_cpu_CO2(cpu_usage, cpu_retrieval_time, cpu.get_cpu_tdp(cpu_info), carbon_intensity)
         print("- In the last {0} seconds your CPU footprint was: {1} g".format(cpu_retrieval_time, co2_emissions))
```
Comment (giulianogiari) [inline src/core/cpuCO2.py]: this is probably not necessary anymore since we have the plot?

File: src/core/gui.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        # define matplotlib font properties
+        font = {
+            'weight': 'normal',
+            'size': 10
+        }
+        matplotlib.rc('font', **font)
+
+        # create a matplotlib line plot structure
+        self.ax = self.canvas.figure.subplots()
+        self.ax.set_xlim([-self.num_x_points, 0])
```
Comment (giulianogiari) [inline src/core/gui.py]: this is not working properly, leaving one space more than expected. not a big deal but can be an easy fix

File: src/core/gui.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        self.set_labels_ticks(x_unit_measurement=self.x_unit_measurement, num_ticks=self.num_x_ticks)
+        self.ax.set_ylabel(f"gCOeq \nevery {self.time_interval}{self.y_unit_measurement}", fontsize="small", rotation="horizontal", horizontalalignment="right")
+        self.ax.set_xlabel("Time")
```
Comment (giulianogiari) [inline src/core/gui.py]: f"Time ({y_unit_measurement})" instead of having the unit in the ticks

File: src/core/gui.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+            'weight': 'normal',
+            'size': 10
+        }
+        matplotlib.rc('font', **font)
+
+        # create a matplotlib line plot structure
+        self.ax = self.canvas.figure.subplots()
+        self.ax.set_xlim([-self.num_x_points, 0])
+        self.set_labels_ticks(x_unit_measurement=self.x_unit_measurement, num_ticks=self.num_x_ticks)
+        self.ax.set_ylabel(f"gCOeq \nevery {self.time_interval}{self.y_unit_measurement}", fontsize="small", rotation="horizontal", horizontalalignment="right")
```
Comment (giulianogiari) [inline src/core/gui.py]: why rotation horizontal?

**Comment by giulianogiari** [comment]
to do: 
• add matplotlib and pyqt5 to requirements.txt
to consider for the future:
• including a cumulative plot, either same figure with two y axes or subplots
• store the data collected to show consumption over days/months (eventually we can make the app run in the background?)

File: src/core/cpuCO2.py (hunk 1 above)
Comment (FrnRmn) [reply src/core/cpuCO2.py]: yep, makes sense

File: src/core/gui.py (hunk 2 above)
Comment (FrnRmn) [reply src/core/gui.py]: nice catch, thanks

File: src/core/gui.py (hunk 4 above)
Comment (FrnRmn) [reply src/core/gui.py]: image
image

I feel like it's more readable, but can be changed

Comment (FrnRmn) [reply src/core/gui.py]: Ok I changed my mind, let's go with vertical :p

**Review by giulianogiari — APPROVED** [review]
good for me

**Comment by FrnRmn** [comment]
@MattiaSilvestri now the problem with the horizontal segments after each data sample should be fixed

File: src/core/cpuCO2.py [hunk 5]
```diff
@@ -26,39 +29,77 @@ def cpu_co2(country_code = None) -> None:
...
 
-        # Combine Co2 with CPU usage
-        print('\nCombining Co2 and CPU data...')
-        co2_emissions = combine_cpu_CO2(cpu_usage, cpu.get_cpu_tdp(cpu_info), carbon_intensity)
-        print("- In the last {0} seconds your CPU footprint was: {1} g".format(cpu_retrieval_time, co2_emissions))
+        # Launch plot
+        app = QApplication(sys.argv)
+        app.setStyleSheet('''
+            QWidget {
+                font-size: 30px;
```
Comment (MattiaSilvestri) [inline src/core/cpuCO2.py:41]: I would add this to a config as well, maybe we can tag with a todo the lines of code we want in the config so we can remember, like: `# TODO: add to config`

File: src/core/cpuCO2.py [hunk 6]
```diff
@@ -26,39 +29,77 @@ def cpu_co2(country_code = None) -> None:
...
+            }
+        ''')
+        myApp = PlotWindowApp(cpu.get_cpu_tdp(cpu_info), carbon_intensity, cpu_retrieval_time, get_interval_emissions, 
+                              num_x_points=200, num_x_ticks=5, x_unit_measurement="m", y_unit_measurement="h")
```
Comment (MattiaSilvestri) [inline src/core/cpuCO2.py]: Add to config.

File: src/core/gui.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        # find y unit of measurement multiplier
+        if y_unit_measurement == "s":
+            self.y_unit_multiplier = 1
+        elif y_unit_measurement == "m":
+            self.y_unit_multiplier = 60
+        elif y_unit_measurement == "h":
+            self.y_unit_multiplier = 3600
+        else:
+            raise ValueError("Please enter a valid unit of measure for the y-axis. Choose between 's', 'm' or 'h'.")
+
```
Comment (MattiaSilvestri) [inline src/core/gui.py:58]: This looks more elegant with a switch-case but it's just aesthetics :)

File: src/core/gui.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        layout.addWidget(self.canvas)
+
+    def insert_ax(self):
+        '''
+        This method populate the empty canvas with a matplotlib axes structure.
+        '''
+        # define matplotlib font properties
+        font = {
+            'weight': 'normal',
+            'size': 10
```
Comment (MattiaSilvestri) [inline src/core/gui.py:86]: Config.

File: src/core/gui.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        # create a matplotlib line plot structure
+        self.ax = self.canvas.figure.subplots()
+        self.ax.set_xlim([-self.num_x_points+1, 0])
+        self.set_labels_ticks(x_unit_measurement=self.x_unit_measurement, num_ticks=self.num_x_ticks)
+        self.ax.set_ylabel(f"gCOeq \nevery {self.time_interval}{self.y_unit_measurement}", fontsize="small", horizontalalignment="right")
+        self.ax.set_xlabel(f"Time ({self.x_unit_measurement})")
+        self.line_plot = None
+        self.ax.set_position([0.1, 0.2, 0.85, 0.7]) #left,bottom,width,height
+        self.ax.grid(color='b', linestyle='-', linewidth=0.1)
+
```
Comment (MattiaSilvestri) [inline src/core/gui.py:99]: Config?

File: src/core/gui.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        This method update the data points to be plotted and display the new plot.
+        '''  
+        # update data to plot
+        self.y = self.y[1:]
+        self.y = np.append(self.y, self.new_value)
+
+        # fill area under the line
+        if self.ax.collections:
+            self.ax.collections[0].remove()
+        self.ax.fill_between(self.x,self.y, 0, color='blue', alpha=.1)
```
Comment (MattiaSilvestri) [inline src/core/gui.py]: Config?

File: src/core/gui.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        
+        # display updated plot
+        if self.line_plot:
+            self.line_plot[0].remove()
+        self.line_plot = self.ax.plot(self.x,self.y, color='#1560BD', alpha=.8)
```
Comment (MattiaSilvestri) [inline src/core/gui.py]: Config for colour?

File: src/core/gui.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        :type x_unit_measurement: str
+        '''  
+        if x_unit_measurement == "s":
+            labels_all = self.x * self.time_interval
+        elif x_unit_measurement == "m":
+            labels_all = self.x * self.time_interval / 60
+        elif x_unit_measurement == "h":
+            labels_all = self.x * self.time_interval / 3600
+        else:
+            raise ValueError("Please enter a valid unit of measure for the x-axis. Choose between 's', 'm' or 'h'.")
```
Comment (MattiaSilvestri) [inline src/core/gui.py]: Same story for switch-case.

**Review by MattiaSilvestri — COMMENTED** [review]
OK in general, if you want you can quickly check my suggestions for adding stuff to config and put some tags if you agree, and merge when you're done.

File: src/core/gui.py (hunk 7 above)
Comment (FrnRmn) [reply src/core/gui.py:58]: I agree! btw match-case for python :)

**Comment by FrnRmn** [comment]
@MattiaSilvestri @giulianogiari updates done

**Review by MattiaSilvestri — APPROVED** [review]

**Pull request merged**